### PR TITLE
try root fix for FP8 tensor

### LINKED ISSUE #143194: [Distributed] Cannot create submesh from submesh error.  Was "is this error message for DeviceMesh needed?"
### 🐛 Describe the bug

```shell
[rank0]: Traceback (most recent call last):
[rank0]:   File "/u/mayank98/scratch/tmp1/dolomite-engine/a.py", line 42, in <module>
[rank0]:     y = model(x)
[rank0]:         ^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1845, in _call_impl
[rank0]:     return inner()
[rank0]:            ^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1772, in inner
[rank0]:     args_kwargs_result = hook(self, args, kwargs)  # type: ignore[misc]
[rank0]:                          ^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_state.py", line 71, in fsdp_hook_wrapper
[rank0]:     return torch._dynamo.disable(func, recursive=True)(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 744, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_state.py", line 239, in _pre_forward
[rank0]:     args, kwargs = self._fsdp_param_group.pre_forward(module, args, kwargs)
[rank0]:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py", line 334, in pre_forward
[rank0]:     self.unshard(self.unshard_async_op)
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py", line 263, in unshard
[rank0]:     self._all_gather_result = foreach_all_gather(
[rank0]:                               ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py", line 139, in foreach_all_gather
[rank0]:     param_all_gather_inputs = _get_param_all_gather_inputs(fsdp_params)
[rank0]:                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py", line 217, in _get_param_all_gather_inputs
[rank0]:     param_all_gather_inputs[i] = fsdp_param.all_gather_inputs
[rank0]:                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param.py", line 705, in all_gather_inputs
[rank0]:     ) = sharded_local_tensor.fsdp_pre_all_gather(self.shard_mesh)
[rank0]:                                                  ^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param.py", line 796, in shard_mesh
[rank0]:     return mesh[mesh.mesh_dim_names[-1]]
[rank0]:            ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/device_mesh.py", line 713, in __getitem__
[rank0]:     slice_mesh_dims = _mesh_resources._get_slice_mesh_dims(
[rank0]:                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/u/mayank98/miniconda3/envs/new-ai/lib/python3.11/site-packages/torch/distributed/device_mesh.py", line 296, in _get_slice_mesh_dims
[rank0]:     raise RuntimeError("Cannot create a submesh from a submesh.")
```

code to replicate:
```python
from torch.distributed._composable.fsdp import fully_shard
from torch.distributed.device_mesh import init_device_mesh
import torch
from torchao.float8 import (
    CastConfig,
    Float8LinearConfig,
    ScalingType,
    convert_to_float8_training,
)


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.l = torch.nn.Linear(2, 2)

    def forward(self, x):
        return self.l(x)
    

mesh = init_device_mesh("cuda", (2, 2, 2), mesh_dim_names=("ddp", "fsdp", "abcd"))

with torch.device(torch.cuda.current_device()):
    model = M()
    x = torch.randn(2, 2, device=torch.cuda.current_device())

model = convert_to_float8_training(
    model,
    config=Float8LinearConfig(
        enable_fsdp_float8_all_gather=True,
        cast_config_input=CastConfig(scaling_type=ScalingType("dynamic")),
        cast_config_weight=CastConfig(scaling_type=ScalingType("dynamic")),
        cast_config_grad_output=CastConfig(scaling_type=ScalingType("dynamic")),
        force_recompute_fp8_weight_in_bwd=True,
    ),
    module_filter_fn=lambda mod, fqn: fqn != "output",
)
model = fully_shard(model, mesh=mesh["ddp", "fsdp"])

y = model(x)
```

issue in torchao: https://github.com/pytorch/ao/issues/1411

### Versions

PyTorch version: 2.6.0.dev20241023+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux 9.4 (Plow) (x86_64)
GCC version: (GCC) 11.4.1 20231218 (Red Hat 11.4.1-3)
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.34

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.14.0-427.42.1.el9_4.x86_64-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 57 bits virtual
Byte Order: Little Endian
CPU(s): 96
On-line CPU(s) list: 0-95
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) Platinum 8468
CPU family: 6
Model: 143
Thread(s) per core: 1
Core(s) per socket: 48
Socket(s): 2
Stepping: 8
CPU(s) scaling MHz: 99%
CPU max MHz: 3800.0000
CPU min MHz: 800.0000
BogoMIPS: 4200.00
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
L1d cache: 4.5 MiB (96 instances)
L1i cache: 3 MiB (96 instances)
L2 cache: 192 MiB (96 instances)
L3 cache: 210 MiB (2 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0-47
NUMA node1 CPU(s): 48-95
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pytorch-triton==3.1.0+cf34004b8a
[pip3] torch==2.6.0.dev20241023+cu124
[pip3] triton==3.0.0
[conda] Could not collect

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### PR DESCRIPTION
**Pull request by mayank31398**
Fixes #143194

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 1]
```diff
@@ -796,6 +797,19 @@ def shard_mesh(self):
...
         raise ValueError(f"Invalid mesh: {mesh}")
 
+    @property
+    def shard_mesh_from_root(self):
+        mesh = self.mesh_info.mesh
+
+        if mesh.ndim == 1:
+            return mesh
+        else:
+            assert mesh.mesh_dim_names is not None
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:807]: `assert mesh.mesh_dim_names is not None` is quite strong. wonder if see such assumption somewhere in FSDP2? I was thinking of fixing this from device mesh api

Comment (awgu) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:807]: @weifengpy I think for HSDP path `mesh.ndim == 2`, we do assume that mesh dim names are specified.

Comment (awgu) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:807]: From this PR: https://github.com/pytorch/pytorch/pull/137436

**Review by fegin — COMMENTED** [review]
I don't think using `_MeshEnv` is a good idea. @wz337 we discussed not to support slicing the submeshes when introducing flattening. We should revisit this option. I don't quite get why FP8 requires this features though. cc., @weifengpy

**Review by fegin — CHANGES_REQUESTED** [review]
I don't think using `_MeshEnv` is a good idea. @wz337 we discussed not to support slicing the submeshes when introducing flattening. We should revisit this option. I don't quite get why FP8 requires this features though. cc., @weifengpy

**Comment by weifengpy** [comment]
> I don't think using `_MeshEnv` is a good idea. @wz337 we discussed not to support slicing the submeshes when introducing flattening. We should revisit this option. I don't quite get why FP8 requires this features though. cc., @weifengpy

fp8 requires all-reduce(abs(max(tensor)) across sharded dim, and avoid all-reduce across replicated dim. so we pass mesh[-1] to fp8 all-gather

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py (hunk 1 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:807]: got you. the main question is whether it's acceptable to use `_MeshEnv.get_root_mesh`. We should have @wz337 tomorrow for discussion

Comment (mayank31398) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:807]: I think it should be acceptable but I could be wrong

**Comment by mayank31398** [comment]
is there any conclusion on how to fix this?

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 2]
```diff
@@ -25,6 +25,7 @@
     FSDPMeshInfo,
     HSDPMeshInfo,
 )
+from torch.distributed.device_mesh import _MeshEnv
```
Comment (fegin) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:28]: Change `_MeshEnv` to `_mesh_resource`

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 3]
```diff
@@ -796,6 +797,19 @@ def shard_mesh(self):
...
+    def shard_mesh_from_root(self):
+        mesh = self.mesh_info.mesh
+
+        if mesh.ndim == 1:
+            return mesh
+        else:
+            assert mesh.mesh_dim_names is not None
+            shard_dim_name = mesh.mesh_dim_names[-1]
+
+            root_mesh = _MeshEnv.get_root_mesh(mesh)
```
Comment (fegin) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:810]: ditto

**Comment by mayank31398** [comment]
@fegin I have made the changes

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 4]
```diff
@@ -25,6 +25,7 @@
     FSDPMeshInfo,
     HSDPMeshInfo,
 )
+from torch.distributed.device_mesh import _mesh_resources
```
[the remark this answers is not in the document]
Comment (mayank31398) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py]: done

**Review by fegin — APPROVED** [review]
Let's land this PR to unblock HSDP + TorchAO. We will need a follow up on DeviceMesh side to discuss how to support submesh slicing. cc., @wz337

**Comment by mayank31398** [comment]
@fegin @awgu can you guys merge this?
would love to get it in today's nightly

**Pull request closed without merging**